compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp

```diff
@@ -578,6 +580,183 @@ static int64_t getMaxVectorTileSize(int64_t numElem, int64_t tileSize,
   return 1;
 }
 
+static bool isInnerMostDimThatMapIsFunctionOf(AffineMap map, int dim) {
+  if (!map.isFunctionOfDim(dim)) {
+    return false;
+  }
+  for (int d = dim + 1; d < map.getNumDims(); ++d) {
+    if (map.isFunctionOfDim(d)) {
+      return false;
+    }
+  }
+  return true;
+}
+
+static int sum(ArrayRef<int64_t> a) {
+  int s = 0;
+  for (int i : a) {
+    s += i;
+  }
+  return s;
+}
+
+// Clamps in-place `vecTileSizes`, ensuring that the resulting vector tile sizes
+// for each opearand of `op` satisfy two requirements:
+// 1. No resulting operand tile size exceeds `eachOperandMaxTileBits`.
+// 2. The sum of all resulting operand tile size does not exceed
+// `allOperandsMaxTileBits`.
+static void limitVectorTileSizes(linalg::LinalgOp op,
+                                 SmallVectorImpl<int64_t> &vecTileSizes,
+                                 int64_t eachOperandMaxTileBits,
+                                 int64_t allOperandsMaxTileBits) {
+  int numLoops = op.getNumLoops();
+  assert(numLoops == vecTileSizes.size());
+  auto indexingMaps = op.getIndexingMapsArray();
+  auto operandTypes = op->getOperandTypes();
+  int numOperands = operandTypes.size();
+
+  SmallVector<int64_t> operandElemBits(numOperands);
+  for (int i : llvm::seq<int>(0, numOperands)) {
+    operandElemBits[i] =
+        cast<ShapedType>(operandTypes[i]).getElementTypeBitWidth();
+  }
```

Comment:
Does below work? It is safer to use `getElementTypeOrSelf` because linalg op accepts taking scalars as input. In the context, the implementation would just crash in debug build because scalars can't be casted to `ShapedType`. Also the `operandTypes` is only used by the loop, so I think we can omit the variable. Instead, we directly use `op->getOperandTypes` here.

```cpp
SmallVector<int64_t> operandElemBits = llvm::map_to_vector(op->getOperandTypes(), getElementTypeOrSelf);
```

Reply:
good catch, thanks!